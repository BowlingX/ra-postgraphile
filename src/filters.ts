import type {
  IntrospectionNamedTypeRef,
  IntrospectionObjectType,
  IntrospectionNonNullTypeRef,
  IntrospectionType,
} from 'graphql'

import type { Filter, FilterMap, FilterSpec, Operator } from './types'

/**
 * Transforms for a certain field type a search value for a field (the key)
 * to a filter that is understood by postgraphile.
 *
 * For example:
 *
 * - type: {kind: "SCALAR", name: "String", ofType: null, __typename: "__Type"}
 * - value: "some keyword"
 * - key: "name"
 *
 * Is transformed to:
 *
 * ```ts
 * {
 *   name: {
 *     includes: "some keyword"
 *   }
 * }
 * ```
 */
export const mapFilterType = (
  type: IntrospectionNamedTypeRef,
  value: any,
  key: string
): Filter | undefined => {
  if (Array.isArray(value)) {
    const spec: FilterSpec = {
      operator: 'in',
      value,
    }

    value = spec
  }

  if (typeof value !== 'object') {
    const typeName = (type?.name ?? '').toLowerCase()

    let operator: Operator = 'equalTo'
    // string uses includes as the default operator for historical reasons
    if (typeName === 'string') {
      operator = 'includes'
    }

    // a type of FullText uses matches as the default operator for historical reasons
    if (typeName === 'fulltext') {
      operator = 'matches'
      value = `${value}:*`
    }

    const spec: FilterSpec = {
      operator,
      value,
    }

    value = spec
  } else {
    // make sure object has a shape of FilterSpec
    if (value?.operator === undefined) {
      throw new Error(`Alternative ${JSON.stringify(value)} filter is not of type FilterSpec`)
    }
  }

  const { operator, value: v } = value

  // react-admin sends the value as undefined when the filter is cleared
  // rather than making every parse function handle that, deal with it here
  if (v === undefined) {
    return undefined
  }

  return {
    [key]: {
      [operator]: v,
    },
  }
}

export const createFilter = (
  fields: { [key: string]: unknown },
  type: IntrospectionType
): FilterMap | undefined => {
  const empty: Filter[] = []

  const filters = Object.keys(fields).reduce((next, key) => {
    const maybeType = (type as IntrospectionObjectType).fields.find((f: any) => f.name === key)
    if (maybeType) {
      const thisType = (maybeType.type as IntrospectionNonNullTypeRef).ofType || maybeType.type

      const filter = mapFilterType(thisType as IntrospectionNamedTypeRef, fields[key], key)
      if (filter === undefined) {
        return next
      }

      return [...next, filter]
    }
    return next
  }, empty)

  if (filters === empty) {
    return undefined
  }

  return { and: filters }
}
