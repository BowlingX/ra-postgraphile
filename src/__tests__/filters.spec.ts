/* tslint:disable:no-expression-statement */
import type { FilterSpec } from '../types'
import { mapFilterType } from '../filters'

describe('filters', () => {
  it('should filter by String', () => {
    expect(mapFilterType({ kind: 'SCALAR', name: 'String' }, 'value', 'id')).toMatchSnapshot()
  })
  it('should filter using a FilterSpec', () => {
    const spec: FilterSpec = {
      operator: 'likeInsensitive',
      value: 'ilike value',
    }

    expect(mapFilterType({ kind: 'SCALAR', name: 'String' }, spec, 'id')).toMatchSnapshot()
  })
  it('should filter by Int', () => {
    expect(mapFilterType({ kind: 'SCALAR', name: 'Int' }, 5, 'id')).toMatchSnapshot()
  })
  it('should filter by BigInt', () => {
    expect(mapFilterType({ kind: 'SCALAR', name: 'Int' }, 5, 'id')).toMatchSnapshot()
  })
  it('should filter by UUID', () => {
    expect(
      mapFilterType({ kind: 'SCALAR', name: 'UUID' }, '02d07429-c2a7-4494-aec9-e8bde9176e86', 'id')
    ).toMatchSnapshot()
  })
  it('should filter by boolean', () => {
    expect(
      mapFilterType({ kind: 'SCALAR', name: 'Boolean' }, true, 'booleanField')
    ).toMatchSnapshot()
  })
  it('should filter with Full Text', () => {
    expect(mapFilterType({ kind: 'SCALAR', name: 'FullText' }, 'test', 'id')).toMatchSnapshot()
  })
  it('should support custom filters via objects', () => {
    const spec: FilterSpec = {
      operator: 'contains',
      value: ['a', 'b'],
    }
    expect(mapFilterType({ kind: 'SCALAR', name: 'String' }, spec, 'id')).toMatchSnapshot()
  })
  it('should support a value of undefined', () => {
    expect(mapFilterType({ kind: 'SCALAR', name: 'String' }, undefined, 'id')).toMatchSnapshot()
  })
  it('should support custom filters with a value of undefined', () => {
    const spec: FilterSpec = {
      operator: 'contains',
      value: undefined,
    }
    expect(mapFilterType({ kind: 'SCALAR', name: 'String' }, spec, 'id')).toMatchSnapshot()
  })
  it('should throw an error if custom filter is not of type FilterSpec', () => {
    expect(() =>
      mapFilterType({ kind: 'SCALAR', name: 'String' }, {}, 'id')
    ).toThrowErrorMatchingSnapshot()

    expect(() =>
      mapFilterType({ kind: 'SCALAR', name: 'String' }, { operator: undefined, value: 'foo' }, 'id')
    ).toThrowErrorMatchingSnapshot()
  })
})
