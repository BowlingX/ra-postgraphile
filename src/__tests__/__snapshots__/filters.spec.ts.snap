// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters should filter by BigInt 1`] = `
Object {
  "id": Object {
    "equalTo": 5,
  },
}
`;

exports[`filters should filter by Int 1`] = `
Object {
  "id": Object {
    "equalTo": 5,
  },
}
`;

exports[`filters should filter by String 1`] = `
Object {
  "id": Object {
    "includes": "value",
  },
}
`;

exports[`filters should filter by UUID 1`] = `
Object {
  "id": Object {
    "equalTo": "02d07429-c2a7-4494-aec9-e8bde9176e86",
  },
}
`;

exports[`filters should filter by boolean 1`] = `
Object {
  "booleanField": Object {
    "equalTo": true,
  },
}
`;

exports[`filters should filter using a FilterSpec 1`] = `
Object {
  "id": Object {
    "likeInsensitive": "ilike value",
  },
}
`;

exports[`filters should filter with Full Text 1`] = `
Object {
  "id": Object {
    "matches": "test:*",
  },
}
`;

exports[`filters should support a value of undefined 1`] = `undefined`;

exports[`filters should support custom filters via objects 1`] = `
Object {
  "id": Object {
    "contains": Array [
      "a",
      "b",
    ],
  },
}
`;

exports[`filters should support custom filters with a value of undefined 1`] = `undefined`;

exports[`filters should throw an error if custom filter is not of type FilterSpec 1`] = `"Alternative {} filter is not of type FilterSpec"`;

exports[`filters should throw an error if custom filter is not of type FilterSpec 2`] = `"Alternative {\\"value\\":\\"foo\\"} filter is not of type FilterSpec"`;
